File: generated_code/cpp/solution_975_1.cpp
#include <vector>
#include <algorithm>

// Definition for a binary tree node if not already defined by the environment.
// Included to ensure self-contained compilability.
#ifndef TREE_NODE_DEFINE
#define TREE_NODE_DEFINE
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
#endif

class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        // Base case: if the node is null, it contributes 0 to the sum.
        if (!root) {
            return 0;
        }
        
        // If the current node's value is strictly less than 'low', 
        // then its entire left subtree will also be less than 'low' due to BST properties.
        // Thus, we only need to search the right subtree.
        if (root->val < low) {
            return rangeSumBST(root->right, low, high);
        }
        
        // If the current node's value is strictly greater than 'high',
        // then its entire right subtree will also be greater than 'high'.
        // Thus, we only need to search the left subtree.
        if (root->val > high) {
            return rangeSumBST(root->left, low, high);
        }
        
        // If the current node's value is within [low, high], it contributes to the sum.
        // We then recursively add the valid range sums from both its left and right subtrees.
        return root->val + rangeSumBST(root->left, low, high) + rangeSumBST(root->right, low, high);
    }
};

File: generated_code/python/solution_975_1.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        """
        Calculates the sum of values of all nodes in a BST within the inclusive range [low, high].
        
        Algorithm: Optimized DFS (Pruning)
        - Since it's a BST, we can skip searching entire subtrees that fall completely out of bounds.
        - If root.val < low, all nodes in the left subtree are also < low, so we only search the right subtree.
        - If root.val > high, all nodes in the right subtree are also > high, so we only search the left subtree.
        - Otherwise, root.val is in range, so we add its value and search both subtrees.
        
        Time Complexity: O(N) in the worst case (where N is the number of nodes), but 
                         substantially faster on average due to pruning.
        Space Complexity: O(H) where H is the height of the tree, representing the recursion stack.
        """
        if not root:
            return 0
        
        # Case 1: Current node value is too small, prune the left subtree.
        if root.val < low:
            return self.rangeSumBST(root.right, low, high)
            
        # Case 2: Current node value is too large, prune the right subtree.
        if root.val > high:
            return self.rangeSumBST(root.left, low, high)
            
        # Case 3: Current node value is within range, include it and check both subtrees.
        return root.val + self.rangeSumBST(root.left, low, high) + self.rangeSumBST(root.right, low, high)